ai-assistants/chat/messages.ts

```diff
@@ -0,0 +1,265 @@
+            },
+          },
+        };
+      },
+    ).filter(notUndefined);
+  });
+};
+
+export interface ProcessorOpts {
+  assistantId: string;
+  instructions: string;
+}
+
+const sleep = (ns: number) => new Promise((resolve) => setTimeout(resolve, ns));
+
+const cache: Record<string, unknown> = {};
+
+const invokeFor = (
+  ctx: AppContext,
+  assistant: AIAssistant,
+  onFunctionCallStart: (
+    tool: RequiredActionFunctionToolCall,
+    props: unknown,
+  ) => void,
+  onFunctionCallEnd: (
+    tool: RequiredActionFunctionToolCall,
+    props: unknown,
+    response: unknown,
+  ) => void,
+) => {
+  return async (call: RequiredActionFunctionToolCall) => {
+    try {
+      const props = JSON.parse(call.function.arguments || "{}");
+
+      const cacheKey = `${call.function.arguments}${call.function.name}`;
+
+      const assistantProps = assistant?.useProps?.(props) ?? props;
+      cache[cacheKey] ??= ctx.invoke(
+        // deno-lint-ignore no-explicit-any
+        call.function.name as any,
+        assistantProps,
+      );
+      onFunctionCallStart(call, assistantProps);
+      const invokeResponse = await cache[cacheKey];
+      onFunctionCallEnd(call, assistantProps, invokeResponse);
+      return {
+        tool_call_id: call.id,
+        output: JSON.stringify(invokeResponse),
+      };
+    } catch (err) {
+      console.error("invoke error", err);
+      return {
+        tool_call_id: call.id,
+        output: "[]",
+      };
+    }
+  };
+};
+/**
+ * Creates a message processor function for the given AI assistant and context.
+ * @param {AIAssistant} assistant - The AI assistant for processing messages.
+ * @param {AppContext} ctx - The application context.
+ * @returns {Promise<(message: ChatMessage) => void>} - A function that processes incoming chat messages.
+ */
+export const messageProcessorFor = async (
+  assistant: AIAssistant,
+  ctx: AppContext,
+) => {
+  const openAI = ctx.openAI;
+  const threads = openAI.beta.threads;
+  const thread = await threads.create();
+  const instructions =
+    `${ctx.instructions}. Introduce yourself as ${assistant.name}. ${assistant.instructions}. Below are arbitrary prompt that gives you information about the current context, it can be empty. \n${
+      (assistant.prompts ?? []).map((prompt) =>
+        `this is the ${prompt.context}: ${prompt.content}`
+      )
+    }. Last, but not least, DO NOT CHANGE THE FUNCTIONS NAMES THAT I'LL GIVE TO YOU, do not remove .ts at the end of function name nor /. If you are positive that your response contains the information that the user requests (like product descriptions, product names, prices, colors, and sizes), add an @ symbol at the end of the phrase. Otherwise, add a # symbol.
+    For example, if the user asks about product availability and you have the information, respond with "The product is in stock. @". If you don't have the information, respond with "I'm sorry, the product is currently unavailable. #".
```

Comment:
```suggestion
    `${ctx.instructions}. Introduce yourself as ${assistant.name}. ${assistant.instructions}. ${assistant.prompts ? "Below are arbitrary prompt that gives you information about the current context, it can be empty." : ""}\n${
      (assistant.prompts ?? []).map((prompt) =>
        `this is the ${prompt.context}: ${prompt.content}`
      )
    }. DO NOT CHANGE FUNCTIONS NAMES, do not remove .ts at the end of function name. do not remove / at the end of function name. If you are positive that your response contains the information that the user requests (like product descriptions, product names, prices, colors, and sizes), add an @ symbol at the end of the phrase. Otherwise, add a # symbol.
    For example, if the user asks about product availability and you have the information, respond with "The product is in stock. @". If you don't have the information, respond with "I'm sorry, the product is currently unavailable. #".
```